panels/updates/cc-updates-panel.c

```diff
@@ -0,0 +1,688 @@
+                                                   self);
+    }
+
+  /* Icon */
+  icon_name = get_network_status_icon_name (self, device, ap);
+  gtk_image_set_from_icon_name (GTK_IMAGE (self->network_status_icon), icon_name, 4);
+
+  /* Name */
+  if (ap)
+    {
+      GBytes *ssid = nm_access_point_get_ssid (ap);
+      g_autofree gchar *title = nm_utils_ssid_to_utf8 (g_bytes_get_data (ssid, NULL), g_bytes_get_size (ssid));
+
+      gtk_label_set_label (GTK_LABEL (self->network_name_label), title);
+    }
+  else
+    {
+      if (connection)
+        gtk_label_set_label (GTK_LABEL (self->network_name_label), _("Connected"));
+      else
+        gtk_label_set_label (GTK_LABEL (self->network_name_label), _("No active connection"));
+    }
+
+  /* Metered status */
+  gtk_widget_set_visible (self->metered_data_label, connection != NULL);
+
+  if (connection)
+    {
+      NMSettingConnection *setting;
+      NMMetered metered;
+
+      setting = nm_connection_get_setting_connection (connection);
+      metered = nm_setting_connection_get_metered (setting);
+
+      if (metered == NM_METERED_YES || metered == NM_METERED_GUESS_YES)
+        gtk_label_set_label (GTK_LABEL (self->metered_data_label), _("Limited data plan connection"));
+      else
+        gtk_label_set_label (GTK_LABEL (self->metered_data_label), _("Unlimited data plan connection"));
+    }
+
+  /* Automatic Updates switch */
+  gtk_widget_set_sensitive (self->automatic_updates_container, connection != NULL);
+
+  ensure_setting_user (self, connection);
+
+  if (connection)
+    {
+      NMSettingUser *setting_user;
+      const gchar *value;
+
+      setting_user = NM_SETTING_USER (nm_connection_get_setting (connection, NM_TYPE_SETTING_USER));
+      value = nm_setting_user_get_data (setting_user, NM_SETTING_ALLOW_DOWNLOADS_WHEN_METERED);
+
+      g_signal_handlers_block_by_func (self->automatic_updates_switch,
+                                       on_automatic_updates_switch_changed_cb,
+                                       self);
+
+      gtk_switch_set_active (GTK_SWITCH (self->automatic_updates_switch), g_strcmp0 (value, "1") == 0);
+
+      g_signal_handlers_unblock_by_func (self->automatic_updates_switch,
+                                         on_automatic_updates_switch_changed_cb,
+                                         self);
+    }
+}
+
+static void
+load_tariff (CcUpdatesPanel *self)
+{
+  g_autoptr(MwtTariffLoader) tariff_loader = NULL;
+  g_autoptr(GInputStream) input_stream = NULL;
+  g_autoptr(GBytes) bytes = NULL;
+  g_autoptr(GFile) tariff_file = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *tariff_path = NULL;
+
+  g_debug ("Loading tariff");
+
+  tariff_path = g_build_filename (g_get_user_config_dir (), SYSTEM_TARIFF_FILENAME, NULL);
+
+  if (!g_file_test (tariff_path, G_FILE_TEST_EXISTS))
+    {
+      g_debug ("Could not open test tariff file");
+      return;
+    }
+
+  if (!g_file_test (tariff_path, G_FILE_TEST_IS_REGULAR))
```

Comment:
No need for these explicit `g_file_test()` calls: just handle the appropriate errors from `g_file_read()`. Otherwise you have a [TOCTTOU race](https://en.wikipedia.org/wiki/Time_of_check_to_time_of_use).